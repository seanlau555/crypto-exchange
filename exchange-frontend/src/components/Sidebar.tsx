import { memo } from 'react'
import { Box } from '@chakra-ui/react'
import { DivEvent } from '../types'
import Item from './Item'

const list = ['BTC', 'ETH', 'SOL', 'DOGE', 'AVAX', 'SAND']
const bases = ['USD']

type Props = {
  onSelect: (value: string) => void
  selectedTicker: string
}

function Sidebar({ selectedTicker, onSelect }: Props) {
  const onClick = (evt: DivEvent) => {
    onSelect(evt.currentTarget.id)
  }
  const currencies = list
    .map((x) => {
      return bases.map((y) => {
        return { current: x, to: y }
      })
    })
    .flat()

  return (
    <>
      <Box
        transition=".4s ease"
        w="300px"
        minW="300px"
        borderColor="gray.100"
        borderRightWidth="1px"
        overflowY="scroll"
        bg="white"
        h="100vh"
      >
        {selectedTicker &&
          currencies.map((x: { current: string; to: string }) => (
            <Item
              key={x.current + x.to}
              current={x.current}
              to={x.to}
              selectedTicker={selectedTicker}
              onClick={onClick}
            />
          ))}
      </Box>
    </>
  )
}

export default memo(Sidebar)
