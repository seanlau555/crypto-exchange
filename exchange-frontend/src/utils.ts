import axios, { AxiosResponse } from 'axios'
import { Bar } from './types'

const Utils = {
  async getAuthToken(oauth_code: string) {
    // returns Authorization Token once we have our OAuth token
    const body = {
      grant_type: 'authorization_code',
      code: oauth_code,
      client_id: import.meta.env.VITE_REACT_APP_CLIENT_ID,
      client_secret: import.meta.env.VITE_REACT_APP_CLIENT_SECRET,
      redirect_uri: import.meta.env.VITE_REACT_APP_REDIRECT_URI,
    }
    // encode data into form encoding
    const encodedBody = Object.keys(body)
      .map((key: string) => `${key}=${encodeURIComponent(body[key])}`)
      .join('&')
    console.log(body)
    // submit POST request
    const response = await axios({
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      method: 'POST',
      url: 'https://api.alpaca.markets/oauth/token',
      data: encodedBody,
    })

    const { data } = response
    window.localStorage.setItem('auth-token', data)
    return data.access_token
  },

  parseResponse(response: AxiosResponse): Array<Bar> | [] {
    const { bars } = response.data
    if (bars) {
      const data = []
      for (let i = 0; i < bars.length; i++) {
        const bar = bars[i]
        const point = {
          date: bar.t,
          open: bar.o,
          low: bar.l,
          high: bar.h,
          close: bar.c,
          volume: bar.v,
        }
        data.push(point)
      }
      return data
    }
    return []
  },
}
export default Utils
