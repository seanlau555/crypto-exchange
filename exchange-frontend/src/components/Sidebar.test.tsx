import { describe, expect, it } from 'vitest'
import Sidebar from './Sidebar'
import { screen, fireEvent } from '@testing-library/react'
import { createWrapper } from '../test/utils'

describe('<Sidebar />', () => {
  it('render sidebar', () => {
    const { container } = createWrapper(
      <Sidebar selectedTicker={'BTCUSD'} onSelect={() => {}} />,
    )
    expect(container).toMatchSnapshot()
  })

  it('onClick item', () => {
    createWrapper(<Sidebar selectedTicker={'BTCUSD'} onSelect={() => {}} />)
    const btnBitcoin = screen.getByTestId('BTCUSD')
    fireEvent.click(btnBitcoin)
    expect(btnBitcoin.id).toBe('BTCUSD')
  })
})
