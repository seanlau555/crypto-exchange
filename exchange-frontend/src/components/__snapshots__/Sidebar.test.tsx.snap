// Vitest Snapshot v1

exports[`<Sidebar /> > render sidebar 1`] = `
<div>
  <div
    class="css-1hn0ibo"
  >
    <div
      class="css-1j9jo68"
      data-testid="BTCUSD"
      id="BTCUSD"
    >
      BTCUSD
      , 
      Price: USD undefined
    </div>
    <div
      class="css-bhtpg9"
      data-testid="ETHUSD"
      id="ETHUSD"
    >
      ETHUSD
      , 
      Price: USD undefined
    </div>
    <div
      class="css-bhtpg9"
      data-testid="SOLUSD"
      id="SOLUSD"
    >
      SOLUSD
      , 
      Price: USD undefined
    </div>
    <div
      class="css-bhtpg9"
      data-testid="DOGEUSD"
      id="DOGEUSD"
    >
      DOGEUSD
      , 
      Price: USD undefined
    </div>
    <div
      class="css-bhtpg9"
      data-testid="AVAXUSD"
      id="AVAXUSD"
    >
      AVAXUSD
      , 
      Price: USD undefined
    </div>
    <div
      class="css-bhtpg9"
      data-testid="SANDUSD"
      id="SANDUSD"
    >
      SANDUSD
      , 
      Price: USD undefined
    </div>
  </div>
</div>
`;
